tensorflow/core/ops/math_ops.cc

```diff
@@ -2365,4 +2365,132 @@ output_max: the computed max output.
 
 )doc");
 
+// --------------------------------------------------------------------------
+
+namespace {
+
+Status ParticalReductionShapeFn(InferenceContext* c) {
+  ShapeHandle handle;
+  DimensionHandle dimhandle;
+  // "indices" must have rank 2 and the number of columns must be 2
+  TF_RETURN_IF_ERROR(c->WithRank(c->input(1), 2, &handle));
+  TF_RETURN_IF_ERROR(c->Merge(c->Dim(c->input(1),1), c->MakeDim(2), &dimhandle));
+  // "data" must have rank at least 1
+  TF_RETURN_IF_ERROR(c->WithRankAtLeast(c->input(0), 1, &handle));
+  // shape of output tensor
+  TF_RETURN_IF_ERROR(c->ReplaceDim(handle, 0, c->Dim(c->input(1),0), &handle));
+  c->set_output(0, handle);
+  return Status::OK();
+}
+
+} // namespace
+
+REGISTER_OP("PartialSum")
+    .Input("data: T")
+    .Input("indices: Tindices")
+    .Output("output: T")
+    .Attr("T: numbertype")
+    .Attr("Tindices: {int32,int64}")
+    .SetShapeFn(ParticalReductionShapeFn)
+    .Doc(R"doc(
+Dynamically sum over the first dimension of a tensor according to start and end
+indices specified at "index". For example if input is
+[[1,1,1],[10,10,10],[100,100,100],[1000,1000,1000]] and index is [[0,1],[1,1],[0,2]],
+the the output will be [[1,1,1],[0,0,0],[11,11,11]]. The data must be at least rank
+one. The indices must be of shape (?,2) where the first column is start indices and
+the second column is end indices. The end index is not included in the sum, which
+means, if you want to do a sum over 0,1,2, the you should have start index 0 and end
+index 3. If end index is smaller than or equal to start, the result will be zero.
+If end index is out of bounds, then the sum will automatically stop at the bound,
+so feel free to put a large number as your end of your index if you want to sum until
+the bound.
+
+data: The source of data where the computation will be taken from.
+indices: start, end indices that controls which part to be included.
+T: the type of data.
+Tindices: the type of indices, must be int32 or int64.
+output: the computed sum values.
+)doc");
+
+REGISTER_OP("PartialProd")
+    .Input("data: T")
+    .Input("indices: Tindices")
+    .Output("output: T")
+    .Attr("T: numbertype")
+    .Attr("Tindices: {int32,int64}")
+    .SetShapeFn(ParticalReductionShapeFn)
+    .Doc(R"doc(
+Dynamically compute the product over the first dimension of a tensor according
+to start and end indices specified at "index". For example if input is
+[[1,2,3],[10,20,30],[100,200,300],[10,10,10]] and index is [[0,1],[1,1],[0,2]],
```

Comment:
Please use an example with different values in each column to make the example more illustrative.